### PR TITLE
Revert noMargin breaking change

### PR DESCRIPTION
**Pull request by robinmetral**
## Purpose

We decided to remove the noMargin prop in the next major version to keep the scope of V3 smaller.

## Approach and changes

- revert the commit that removed the noMargin
  - addressed merge conflicts (new logger util and more) so this. is unfortunately not a clean revert, please review as any normal PR
- remove relevant changelog entry
 
## Definition of done

* [ ] Development completed
* [ ] Reviewers assigned
* [ ] Unit and integration tests
* [ ] Meets minimum browser support
* [ ] Meets accessibility requirements

### CODE REVIEW

File: packages/circuit-ui/components/Carousel/__snapshots__/Carousel.spec.js.snap
Comment (robinmetral) [inline packages/circuit-ui/components/Carousel/__snapshots__/Carousel.spec.js.snap:202]: Hm, should the carousel components also be. given the noMargin prop? Was this an unintended visual change?

File: packages/circuit-ui/components/Checkbox/Checkbox.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/Checkbox/Checkbox.tsx:42]: Yes, it's technically an input, but still I think it would make more sense to call it checkbox here

```suggestion
   * Removes the default bottom margin from the checkbox.
```

File: packages/circuit-ui/components/Input/Input.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/Input/Input.tsx:127]: Oops, this was removed in the context of requiring the label prop and should not be reintroduced

File: packages/circuit-ui/components/List/List.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/List/List.tsx:34]: This shouldn't change

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/Selector/Selector.tsx:63]: selector

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/Selector/Selector.tsx:78]: The default Selector margin should be in baseStyles

File: packages/circuit-ui/components/SubHeadline/SubHeadline.spec.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/SubHeadline/SubHeadline.spec.tsx:34]: This was apparently a leftover test, we'll be able to remove it next time

File: packages/circuit-ui/components/Toggle/Toggle.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/Toggle/Toggle.tsx:37]: toggle

**Review by robinmetral — COMMENTED** [review]
I found quite a few leftovers from the merge conflicts, will address now

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (robinmetral) [reply packages/circuit-ui/components/Selector/Selector.tsx:78]: Fixed

File: packages/circuit-ui/components/Carousel/__snapshots__/Carousel.spec.js.snap
Comment (robinmetral) [reply packages/circuit-ui/components/Carousel/__snapshots__/Carousel.spec.js.snap:202]: As per Chromatic, it looks like it's not actually changing anything

File: packages/circuit-ui/components/Anchor/__snapshots__/Anchor.spec.tsx.snap
Comment (connor-baer) [reply packages/circuit-ui/components/Anchor/__snapshots__/Anchor.spec.tsx.snap:9]: Yes, it was. The Anchor extends the Body component which has the built-in spacing, however, when introducing the Anchor component, we were already planning to remove the built-in spacing, so I decided to exclude it from the Anchor component from the start.

File: packages/circuit-ui/components/Carousel/__snapshots__/Carousel.spec.js.snap
Comment (connor-baer) [reply packages/circuit-ui/components/Carousel/__snapshots__/Carousel.spec.js.snap:202]: The margin is reset a few lines down.

**Review by connor-baer — APPROVED** [review]

**Review by amelako — APPROVED** [review]

File: packages/circuit-ui/components/List/List.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/List/List.tsx:101]: @connor-baer WDYT? The rest are smaller visual improvements in the stories, after you review this part I think we can merge this 🙂

Comment (connor-baer) [reply packages/circuit-ui/components/List/List.tsx:101]: How does this affect nested lists?

Comment (robinmetral) [reply packages/circuit-ui/components/List/List.tsx:101]: Do you mean like [this](https://oss-circuit-ui-ai0uf0bh1-sumup.vercel.app/?path=/story/typography-list--nested)? I think it's fine

Comment (connor-baer) [reply packages/circuit-ui/components/List/List.tsx:101]: The problem is when there's another list item after the nested list:

Screenshot 2021-07-13 at 18 23 07

Comment (robinmetral) [reply packages/circuit-ui/components/List/List.tsx:101]: 👀

Comment (robinmetral) [reply packages/circuit-ui/components/List/List.tsx:101]: Nice catch, but this is a bug on `main` 🐛 

If you add `noMargin` to the nested list (which is the expected behavior, fixed the story) it would look like this:

Screenshot 2021-07-13 at 18 57 35

- there is indeed an issue in the nested list because it doesn't have the 8px spacing with the next item 🐛 
- the missing spacing with the paragraph below the list is what we want, we can style it with the spacing mixin

An easy alternative would be _not_ to reset the outer spacing:

Screenshot 2021-07-13 at 19 00 43

- the list itself looks good
- but there is outer spacing, we don't want this optimally (though we had it before even with noMargin passed)

We can also address the actual bug but it looks pretty tricky to me, because the nested list isn't aware of any parent. Do you have a CSS magic trick for this. or should we keep it for a 3.x.x bugfix?

Comment (connor-baer) [reply packages/circuit-ui/components/List/List.tsx:101]: In the future, we can use the `:not()` selector for this: https://twitter.com/leaverou/status/1354760561087676416. Right now, I'm not sure. 

This is an edge case and since this is an existing bug, let's address this in a separate PR 👍🏻

**Review by connor-baer — APPROVED** [review]
🙌🏻

File: packages/circuit-ui/components/List/List.tsx
Comment (robinmetral) [reply packages/circuit-ui/components/List/List.tsx:102]: This wasn't CSS magic, it was CSS 101 🤦 (I blame Tuesdays)

6fcc26c

Comment (robinmetral) [reply packages/circuit-ui/components/List/List.tsx:102]: Ah but no, this also doesn't work since if a list _ends_ with a nested item, we'll be back at the unwanted outer margin

Comment (robinmetral) [reply packages/circuit-ui/components/List/List.tsx]: I think I've got it now, we can reset a nested list's bottom margin when it's the `:last-child`

4ddb9b6

**Comment by robinmetral** [comment]
@connor-baer I won't merge this today after the last changes, please take a quick look when you have time tomorrow 🙏 I think the list bug is also solved by this

File: packages/circuit-ui/components/List/List.tsx
Comment (connor-baer) [reply packages/circuit-ui/components/List/List.tsx:102]: You're the best 👏🏻

**Review by connor-baer — APPROVED** [review]

**Pull request merged**
